compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp

```diff
@@ -31,18 +31,55 @@ namespace IREE {
 namespace Flow {
 namespace {
 
+static int64_t costOfDomain(ArrayRef<int64_t> domain) {
+  int64_t product = 1;
+  for (int64_t size : domain) {
+    if (size == mlir::ShapedType::kDynamic) return INT64_MAX;
+    product *= size;
+  }
+  return product;
+};
+
 // Estimates the evaluation cost of a linalg op using a heuristic cost model.
 static int64_t estimateLinalgOpCost(linalg::LinalgOp op) {
-  if (op.hasDynamicShape()) {
-    // Note: bounded dynamic shapes would be interesting, if the compiler used
-    // them. For now just treat dynamic shapes as arbitrarily large.
-    return INT64_MAX;
-  }
+  // For linalg ops we know the iteration domain, so return the number
+  // of iterations of the iteration domain (or INT64_MAX for dynamic.)
+  int64_t cost = costOfDomain(op.getStaticLoopRanges());
+  LLVM_DEBUG(llvm::dbgs() << "// " << op->getName() << " cost: " << cost
+                          << "\n");
+  return cost;
+}
 
-  int64_t cost = 1;
-  for (auto loopRange : op.getStaticLoopRanges()) {
-    cost *= loopRange;
+static TensorType getMainTensorForLinalgExtOp(Operation *op) {
+  TensorType main;
+  auto operandTypes = llvm::to_vector(op->getOperandTypes());
+  auto resultTypes = llvm::to_vector(op->getResultTypes());
+  for (Type t : llvm::concat<Type>(operandTypes, resultTypes)) {
+    if (auto tensorType = t.dyn_cast<TensorType>()) {
+      if (!main) {
+        main = tensorType;
+      } else if (costOfDomain(tensorType.getShape()) >
+                 costOfDomain(main.getShape())) {
```

Comment:
maybe cache the `costOfDomain(main.getShape())`, so we can have less computation?

Reply:
I don't think it's significant enough to uglify the code.

Reply:
That's true.. I think the LinalgExt ops are not even fusing with others in current pipeline, except set/unset_encodings.